src/Shared/VolumeNameGenerator.cs

```diff
@@ -0,0 +1,52 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Aspire.Hosting.ApplicationModel;
+
+namespace Aspire.Hosting.Utils;
+
+internal static class VolumeNameGenerator
+{
+    /// <summary>
+    /// Creates a volume name with the form <c>$"{applicationName}-{resourceName}-{suffix}</c>, e.g. <c>"myapplication-postgres-data"</c>.
+    /// </summary>
+    /// <remarks>
+    /// If the application name contains chars that are invalid for a volume name, the prefix <c>"volume-"</c> will be used instead.
+    /// </remarks>
+    public static string CreateVolumeName<T>(IResourceBuilder<T> builder, string suffix) where T : IResource
+    {
+        if (!HasOnlyValidChars(suffix))
+        {
+            throw new ArgumentException($"The suffix '{suffix}' contains invalid characters. Only [a-zA-Z0-9_.-] are allowed.", nameof(suffix));
+        }
+
+        // Create volume name like "myapplication-postgres-data"
+        var applicationName = builder.ApplicationBuilder.Environment.ApplicationName;
+        var resourceName = builder.Resource.Name;
+        return $"{(HasOnlyValidChars(applicationName) ? applicationName : "volume")}-{resourceName}-{suffix}";
```

Comment:
If the `applicationName` has invalid chars then should it be sanitized (like replace invalid ones with `_`) and used instead of defaulting to `volume`?
Does a user need to be able to reliably predict this name? Just wondering because here we silently switch to `"volume"` if `applicationName` has invalid chars.

Reply:
I started down the path of sanitization but ended up opting for this instead for simplicity sake. I think it'll be rare that an application name contains invalid chars so the extra complexity of char substitution (including dealing with what char to substitute if it's the first character that's invalid) didn't seem worth it.